File: app/src/main/java/com/sameetasadullah/i180479_180531/TuluEncryption.java
package com.sameetasadullah.i180479_180531;

import android.util.Base64;
import android.util.Log;
import java.util.HashMap;
import java.util.Map;


public class TuluEncryption {
    
    private static final String TAG = "TuluEncryption";
    
    // Tulu/Kannada character mapping for common syllables
    private static final Map<String, String> TULU_MAPPING = new HashMap<String, String>() {{
        put("a", "ಅ"); put("e", "ಎ"); put("i", "ಇ"); put("o", "ಒ"); put("u", "ಉ");
        put("ka", "ಕ"); put("ga", "ಗ"); put("cha", "ಚ"); put("ja", "ಜ"); put("ta", "ಟ");
        put("da", "ಡ"); put("na", "ನ"); put("pa", "ಪ"); put("ba", "ಬ"); put("ma", "ಮ");
        put("ya", "ಯ"); put("ra", "ರ"); put("la", "ಲ"); put("va", "ವ"); put("sa", "ಸ"); put("ha", "ಹ");
    }};
    
    // Reverse mapping for decryption
    private static final Map<String, String> REVERSE_TULU_MAPPING = new HashMap<>();
    
    // Number to symbol mapping
    private static final Map<String, String> NUMBER_MAPPING = new HashMap<String, String>() {{
        put("0", "⌖"); put("1", "⚙"); put("2", "☉"); put("3", "✪"); put("4", "⦿");
        put("5", "☬"); put("6", "⛁"); put("7", "♜"); put("8", "✩"); put("9", "⚶");
    }};
    private static final Map<String, String> REVERSE_NUMBER_MAPPING = new HashMap<>();
    
    // Symbol to special character mapping
    private static final Map<String, String> SYMBOL_MAPPING = new HashMap<String, String>() {{
        put(",", "⌁"); put(".", "⍕"); put("?", "⌬"); put("!", "✶"); put("-", "⛊");
        put("_", "⍟"); put(":", "⛎"); put(";", "✺"); put("(", "⏢"); put(")", "⏣");
        put(" ", "◦"); put("@", "⚡"); put("#", "⊗"); put("$", "⌘"); put("%", "◈");
        put("&", "⟐"); put("*", "✦"); put("+", "⊕"); put("=", "⚌"); put("/", "⟋");
    }};
    private static final Map<String, String> REVERSE_SYMBOL_MAPPING = new HashMap<>();
    
    // Golden ratio for mathematical scrambling
    private static final double PHI = (1 + Math.sqrt(5)) / 2;
    
    // Encryption marker to identify Tulu-encrypted messages
    private static final String TULU_MARKER = "ತುಳು:"; // "Tulu:" in Kannada
    
    // Initialize reverse mappings
    static {
        // Initialize reverse mappings
        for (Map.Entry<String, String> entry : TULU_MAPPING.entrySet()) {
            REVERSE_TULU_MAPPING.put(entry.getValue(), entry.getKey());
        }
        for (Map.Entry<String, String> entry : NUMBER_MAPPING.entrySet()) {
            REVERSE_NUMBER_MAPPING.put(entry.getValue(), entry.getKey());
        }
        for (Map.Entry<String, String> entry : SYMBOL_MAPPING.entrySet()) {
            REVERSE_SYMBOL_MAPPING.put(entry.getValue(), entry.getKey());
        }
    }
    

    private String scrambleText(String text) {
        try {
            StringBuilder result = new StringBuilder();
            for (char c : text.toCharArray()) {
                int transformed = (int)((int)(c * PHI) % 500) + 1000;
                result.append((char)transformed);
            }
            return result.toString();
        } catch (Exception e) {
            Log.e(TAG, "Error in scrambleText: " + e.getMessage());
            return text; // Fallback to original
        }
    }
    

    private String unscrambleText(String scrambledText) {
        try {
            StringBuilder result = new StringBuilder();
            for (char c : scrambledText.toCharArray()) {
                int target = (int) c;
                boolean found = false;
                
                // Try to find the original character
                for (int code = 32; code < 127; code++) { // Printable ASCII range
                    if ((int)(code * PHI) % 500 + 1000 == target) {
                        result.append((char)code);
                        found = true;
                        break;
                    }
                }
                if (!found) {
                    result.append("?"); // Placeholder for unrecoverable characters
                }
            }
            return result.toString();
        } catch (Exception e) {
            Log.e(TAG, "Error in unscrambleText: " + e.getMessage());
            return scrambledText; // Fallback
        }
    }
    

    public String encryptMessage(String message) {
        try {
            if (message == null || message.isEmpty()) {
                return message;
            }
            
            StringBuilder result = new StringBuilder();
            String messageLower = message.toLowerCase();
            int i = 0;
            
            while (i < message.length()) {
                boolean charFound = false;
                
                // Check for two-character mappings first (like "ka", "ga")
                if (i < message.length() - 1) {
                    String twoChar = messageLower.substring(i, i + 2);
                    if (TULU_MAPPING.containsKey(twoChar)) {
                        result.append(TULU_MAPPING.get(twoChar));
                        i += 2;
                        charFound = true;
                        continue;
                    }
                }
                
                // Check single character mappings
                String singleChar = String.valueOf(messageLower.charAt(i));
                if (TULU_MAPPING.containsKey(singleChar)) {
                    result.append(TULU_MAPPING.get(singleChar));
                    charFound = true;
                } else if (NUMBER_MAPPING.containsKey(singleChar)) {
                    result.append(NUMBER_MAPPING.get(singleChar));
                    charFound = true;
                } else if (SYMBOL_MAPPING.containsKey(singleChar)) {
                    result.append(SYMBOL_MAPPING.get(singleChar));
                    charFound = true;
                }
                
                // If no mapping found, use golden ratio scrambling
                if (!charFound) {
                    result.append(scrambleText(String.valueOf(message.charAt(i)))); // Preserve case for scrambling
                }
                
                i++;
            }
            
            // Add Tulu marker prefix for identification
            String encryptedResult = TULU_MARKER + result.toString();
            
            // Base64 encode for safe storage/transmission
            String b64Result = Base64.encodeToString(encryptedResult.getBytes("UTF-8"), Base64.URL_SAFE | Base64.NO_WRAP);
            
            Log.d(TAG, "Tulu encryption successful");
            return b64Result;
            
        } catch (Exception e) {
            Log.e(TAG, "Tulu encryption error: " + e.getMessage());
            return message; // Fallback to original message
        }
    }
    

    public String decryptMessage(String encryptedMessage) {
        try {
            if (encryptedMessage == null || encryptedMessage.isEmpty()) {
                return encryptedMessage;
            }
            
            // Check if this is a Tulu-encrypted message
            if (!isTuluEncrypted(encryptedMessage)) {
                return encryptedMessage; // Not Tulu-encrypted, return as-is
            }
            
            // Base64 decode
            byte[] decodedBytes = Base64.decode(encryptedMessage, Base64.URL_SAFE);
            String decodedMessage = new String(decodedBytes, "UTF-8");
            
            // Remove Tulu marker
            String textToDecrypt = decodedMessage;
            if (decodedMessage.startsWith(TULU_MARKER)) {
                textToDecrypt = decodedMessage.substring(TULU_MARKER.length());
            }
            
            StringBuilder result = new StringBuilder();
            for (int i = 0; i < textToDecrypt.length(); i++) {
                String singleChar = String.valueOf(textToDecrypt.charAt(i));
                boolean charFound = false;
                
                // Check reverse mappings
                if (REVERSE_NUMBER_MAPPING.containsKey(singleChar)) {
                    result.append(REVERSE_NUMBER_MAPPING.get(singleChar));
                    charFound = true;
                } else if (REVERSE_SYMBOL_MAPPING.containsKey(singleChar)) {
                    result.append(REVERSE_SYMBOL_MAPPING.get(singleChar));
                    charFound = true;
                } else if (REVERSE_TULU_MAPPING.containsKey(singleChar)) {
                    result.append(REVERSE_TULU_MAPPING.get(singleChar));
                    charFound = true;
                }
                
                // If no mapping found, it's a scrambled character
                if (!charFound) {
                    result.append(unscrambleText(singleChar));
                }
            }
            
            Log.d(TAG, "Tulu decryption successful");
            return result.toString();
            
        } catch (Exception e) {
            Log.e(TAG, "Tulu decryption error: " + e.getMessage());
            return encryptedMessage; // Fallback to original
        }
    }
    

    public boolean isTuluEncrypted(String message) {
        try {
            if (message == null || message.length() < 10) {
                return false;
            }
            
            // Try to base64 decode
            byte[] decodedBytes = Base64.decode(message, Base64.URL_SAFE);
            String decoded = new String(decodedBytes, "UTF-8");
            
            // Check for Tulu marker
            return decoded.startsWith(TULU_MARKER);
            
        } catch (Exception e) {
            return false;
        }
    }
    

    public static class HybridEncryption {
        private final TuluEncryption tuluEncryption;
        private final boolean useTulu;
        private final boolean useAes;
        
        public HybridEncryption(boolean useTulu, boolean useAes) {
            this.useTulu = useTulu;
            this.useAes = useAes;
            this.tuluEncryption = useTulu ? new TuluEncryption() : null;
        }
        
        public HybridEncryption() {
            this(true, true); // Default: use both layers
        }
        

        public String encryptMessage(String message) {
            try {
                String encryptedMessage = message;
                
                // First layer: Tulu encryption (character obfuscation)
                if (useTulu && tuluEncryption != null) {
                    encryptedMessage = tuluEncryption.encryptMessage(encryptedMessage);
                    Log.d(TAG, "Applied Tulu encryption layer");
                }
                
                // Second layer: Could add AES encryption here if needed
                // For now, we'll keep it simple with just Tulu
                
                return encryptedMessage;
                
            } catch (Exception e) {
                Log.e(TAG, "Hybrid encryption error: " + e.getMessage());
                return message;
            }
        }
        

        public String decryptMessage(String encryptedMessage) {
            try {
                String decryptedMessage = encryptedMessage;
                
                // Reverse order: AES decryption first (if implemented), then Tulu
                
                // Tulu decryption
                if (useTulu && tuluEncryption != null) {
                    decryptedMessage = tuluEncryption.decryptMessage(decryptedMessage);
                    Log.d(TAG, "Applied Tulu decryption layer");
                }
                
                return decryptedMessage;
                
            } catch (Exception e) {
                Log.e(TAG, "Hybrid decryption error: " + e.getMessage());
                return encryptedMessage;
            }
        }
        
        /**
         * Check if message is encrypted by this hybrid system
         * @param message Message to check
         * @return True if encrypted
         */
        public boolean isHybridEncrypted(String message) {
            return tuluEncryption != null && tuluEncryption.isTuluEncrypted(message);
        }
    }
    
    /**
     * Utility class for easy access to Tulu encryption
     */
    public static class Helper {
        private static TuluEncryption instance;
        private static HybridEncryption hybridInstance;
        
        public static synchronized TuluEncryption getInstance() {
            if (instance == null) {
                instance = new TuluEncryption();
            }
            return instance;
        }
        
        public static synchronized HybridEncryption getHybridInstance() {
            if (hybridInstance == null) {
                hybridInstance = new HybridEncryption();
            }
            return hybridInstance;
        }
        
        // Convenience methods
        public static String encrypt(String message) {
            return getInstance().encryptMessage(message);
        }
        
        public static String decrypt(String encryptedMessage) {
            return getInstance().decryptMessage(encryptedMessage);
        }
        
        public static String hybridEncrypt(String message) {
            return getHybridInstance().encryptMessage(message);
        }
        
        public static String hybridDecrypt(String encryptedMessage) {
            return getHybridInstance().decryptMessage(encryptedMessage);
        }
        
        public static boolean isEncrypted(String message) {
            return getInstance().isTuluEncrypted(message);
        }
    }
    
    /**
     * Demo method to show character mapping
     * @param text Text to demonstrate mapping for
     * @return Mapping demonstration string
     */
    public static String demonstrateMapping(String text) {
        StringBuilder demo = new StringBuilder();
        demo.append("Tulu Character Mapping Demo:\n");
        demo.append("Original: '").append(text).append("'\n");
        demo.append("Character mappings:\n");
        
        String textLower = text.toLowerCase();
        int i = 0;
        while (i < text.length()) {
            // Check two-character mappings first
            if (i < text.length() - 1) {
                String twoChar = textLower.substring(i, i + 2);
                if (TULU_MAPPING.containsKey(twoChar)) {
                    demo.append("  '").append(twoChar).append("' → '").append(TULU_MAPPING.get(twoChar)).append("'\n");
                    i += 2;
                    continue;
                }
            }
            
            // Check single character mappings
            String singleChar = String.valueOf(textLower.charAt(i));
            if (TULU_MAPPING.containsKey(singleChar)) {
                demo.append("  '").append(singleChar).append("' → '").append(TULU_MAPPING.get(singleChar)).append("'\n");
            } else if (NUMBER_MAPPING.containsKey(singleChar)) {
                demo.append("  '").append(singleChar).append("' → '").append(NUMBER_MAPPING.get(singleChar)).append("'\n");
            } else if (SYMBOL_MAPPING.containsKey(singleChar)) {
                demo.append("  '").append(singleChar).append("' → '").append(SYMBOL_MAPPING.get(singleChar)).append("'\n");
            } else {
                demo.append("  '").append(text.charAt(i)).append("' → (scrambled)\n");
            }
            i++;
        }
        
        return demo.toString();
    }
}

File: server/tulu_encryption.py
#!/usr/bin/env python3
"""
Tulu Language Based Encryption Module for Bisto Chat
Advanced encryption using Kannada/Tulu characters, symbol mapping, and golden ratio scrambling
"""

import os
import math
import base64
import logging
from datetime import datetime
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

class TuluEncryption:
    """
    Advanced Tulu-based encryption system using:
    - Kannada/Tulu character mapping for common syllables
    - Symbol mapping for punctuation and numbers
    - Golden ratio scrambling for unknown characters
    - Multiple layers of obfuscation
    """
    
    def __init__(self):
        """Initialize Tulu encryption with character mappings"""
        
        # Tulu/Kannada character mapping for common syllables
        self.tulu_mapping = {
            "a": "ಅ", "e": "ಎ", "i": "ಇ", "o": "ಒ", "u": "ಉ",
            "ka": "ಕ", "ga": "ಗ", "cha": "ಚ", "ja": "ಜ", "ta": "ಟ",
            "da": "ಡ", "na": "ನ", "pa": "ಪ", "ba": "ಬ", "ma": "ಮ",
            "ya": "ಯ", "ra": "ರ", "la": "ಲ", "va": "ವ", "sa": "ಸ", "ha": "ಹ"
        }
        
        # Reverse mapping for decryption
        self.reverse_tulu_mapping = {v: k for k, v in self.tulu_mapping.items()}
        
        # Number to symbol mapping
        self.number_mapping = {
            "0": "⌖", "1": "⚙", "2": "☉", "3": "✪", "4": "⦿",
            "5": "☬", "6": "⛁", "7": "♜", "8": "✩", "9": "⚶"
        }
        self.reverse_number_mapping = {v: k for k, v in self.number_mapping.items()}
        
        # Symbol to special character mapping
        self.symbol_mapping = {
            ",": "⌁", ".": "⍕", "?": "⌬", "!": "✶", "-": "⛊",
            "_": "⍟", ":": "⛎", ";": "✺", "(": "⏢", ")": "⏣",
            " ": "◦", "@": "⚡", "#": "⊗", "$": "⌘", "%": "◈",
            "&": "⟐", "*": "✦", "+": "⊕", "=": "⚌", "/": "⟋"
        }
        self.reverse_symbol_mapping = {v: k for k, v in self.symbol_mapping.items()}
        
        # Golden ratio for mathematical scrambling
        self.phi = (1 + math.sqrt(5)) / 2
        
        # Encryption marker to identify Tulu-encrypted messages
        self.tulu_marker = "ತುಳು:"  # "Tulu:" in Kannada
        
    def scramble_text(self, text: str) -> str:
        """
        Scramble text using golden ratio mathematical transformation
        
        Args:
            text (str): Text to scramble
            
        Returns:
            str: Scrambled text using Unicode characters
        """
        try:
            return ''.join(chr((int(ord(c) * self.phi) % 500) + 1000) for c in text)
        except Exception as e:
            logger.error(f"Error in scramble_text: {e}")
            return text  # Fallback to original
            
    def unscramble_text(self, scrambled_text: str) -> str:
        """
        Unscramble text by reversing golden ratio transformation
        
        Args:
            scrambled_text (str): Scrambled text to reverse
            
        Returns:
            str: Original text
        """
        try:
            result = ""
            for char in scrambled_text:
                target = ord(char)
                found = False
                # Try to find the original character
                for code in range(32, 127):  # Printable ASCII range
                    if (int(code * self.phi) % 500) + 1000 == target:
                        result += chr(code)
                        found = True
                        break
                if not found:
                    result += "?"  # Placeholder for unrecoverable characters
            return result
        except Exception as e:
            logger.error(f"Error in unscramble_text: {e}")
            return scrambled_text  # Fallback
    
    def encrypt_message(self, message: str) -> str:
        """
        Encrypt a message using Tulu character mapping and golden ratio scrambling
        
        Args:
            message (str): Plain text message to encrypt
            
        Returns:
            str: Tulu-encrypted message with marker
        """
        try:
            if not message:
                return message
                
            result = ""
            i = 0
            message_lower = message.lower()
            
            while i < len(message):
                char_found = False
                
                # Check for two-character mappings first (like "ka", "ga")
                if i < len(message) - 1:
                    two_char = message_lower[i:i+2]
                    if two_char in self.tulu_mapping:
                        result += self.tulu_mapping[two_char]
                        i += 2
                        char_found = True
                        continue
                
                # Check single character mappings
                char = message_lower[i]
                if char in self.tulu_mapping:
                    result += self.tulu_mapping[char]
                    char_found = True
                elif char in self.number_mapping:
                    result += self.number_mapping[char]
                    char_found = True
                elif char in self.symbol_mapping:
                    result += self.symbol_mapping[char]
                    char_found = True
                
                # If no mapping found, use golden ratio scrambling
                if not char_found:
                    result += self.scramble_text(message[i])  # Preserve case for scrambling
                
                i += 1
            
            # Add Tulu marker prefix for identification
            encrypted_result = self.tulu_marker + result
            
            # Base64 encode for safe storage/transmission
            b64_result = base64.urlsafe_b64encode(encrypted_result.encode('utf-8')).decode('utf-8')
            
            logger.debug(f"Tulu encryption successful")
            return b64_result
            
        except Exception as e:
            logger.error(f"Tulu encryption error: {e}")
            return message  # Fallback to original message
    
    def decrypt_message(self, encrypted_message: str) -> str:
        """
        Decrypt a Tulu-encrypted message
        
        Args:
            encrypted_message (str): Tulu-encrypted message
            
        Returns:
            str: Decrypted plain text message
        """
        try:
            if not encrypted_message:
                return encrypted_message
            
            # Check if this is a Tulu-encrypted message
            if not self.is_tulu_encrypted(encrypted_message):
                return encrypted_message  # Not Tulu-encrypted, return as-is
            
            # Base64 decode
            decoded_message = base64.urlsafe_b64decode(encrypted_message.encode('utf-8')).decode('utf-8')
            
            # Remove Tulu marker
            if decoded_message.startswith(self.tulu_marker):
                text_to_decrypt = decoded_message[len(self.tulu_marker):]
            else:
                text_to_decrypt = decoded_message
            
            result = ""
            for char in text_to_decrypt:
                char_found = False
                
                # Check reverse mappings
                if char in self.reverse_number_mapping:
                    result += self.reverse_number_mapping[char]
                    char_found = True
                elif char in self.reverse_symbol_mapping:
                    result += self.reverse_symbol_mapping[char]
                    char_found = True
                elif char in self.reverse_tulu_mapping:
                    result += self.reverse_tulu_mapping[char]
                    char_found = True
                
                # If no mapping found, it's a scrambled character
                if not char_found:
                    result += self.unscramble_text(char)
            
            logger.debug("Tulu decryption successful")
            return result
            
        except Exception as e:
            logger.error(f"Tulu decryption error: {e}")
            return encrypted_message  # Fallback to original
    
    def is_tulu_encrypted(self, message: str) -> bool:
        """
        Check if a message is Tulu-encrypted
        
        Args:
            message (str): Message to check
            
        Returns:
            bool: True if message appears to be Tulu-encrypted
        """
        try:
            if not message or len(message) < 10:
                return False
            
            # Try to base64 decode
            decoded = base64.urlsafe_b64decode(message.encode('utf-8')).decode('utf-8')
            
            # Check for Tulu marker
            return decoded.startswith(self.tulu_marker)
            
        except Exception:
            return False
    
    def encrypt_with_metadata(self, message: str, sender_uid: str = None) -> Dict[str, Any]:
        """
        Encrypt message with additional metadata
        
        Args:
            message (str): Message to encrypt
            sender_uid (str, optional): Sender's unique identifier
            
        Returns:
            dict: Encrypted message with metadata
        """
        try:
            encrypted_message = self.encrypt_message(message)
            
            return {
                'encrypted_message': encrypted_message,
                'encryption_type': 'tulu',
                'encryption_timestamp': datetime.now().isoformat(),
                'sender_uid': sender_uid,
                'message_length': len(message),
                'encrypted_length': len(encrypted_message)
            }
            
        except Exception as e:
            logger.error(f"Error in encrypt_with_metadata: {e}")
            return {
                'encrypted_message': message,
                'encryption_type': 'none',
                'error': str(e)
            }

# Enhanced encryption wrapper that combines AES and Tulu
class HybridEncryption:
    """
    Hybrid encryption system combining AES-256 and Tulu encryption
    """
    
    def __init__(self, use_tulu: bool = True, use_aes: bool = True):
        """
        Initialize hybrid encryption
        
        Args:
            use_tulu (bool): Enable Tulu encryption layer
            use_aes (bool): Enable AES encryption layer
        """
        self.use_tulu = use_tulu
        self.use_aes = use_aes
        
        if use_tulu:
            self.tulu_encryptor = TuluEncryption()
        
        if use_aes:
            # Import AES encryption from existing module
            try:
                from encryption_utils import MessageEncryption
                self.aes_encryptor = MessageEncryption()
            except ImportError:
                logger.warning("AES encryption not available")
                self.use_aes = False
    
    def encrypt_message(self, message: str) -> str:
        """
        Encrypt message using hybrid approach
        
        Args:
            message (str): Plain text message
            
        Returns:
            str: Encrypted message
        """
        try:
            encrypted_message = message
            
            # First layer: Tulu encryption (character obfuscation)
            if self.use_tulu:
                encrypted_message = self.tulu_encryptor.encrypt_message(encrypted_message)
                logger.debug("Applied Tulu encryption layer")
            
            # Second layer: AES encryption (strong cryptographic security)
            if self.use_aes and hasattr(self, 'aes_encryptor'):
                encrypted_message = self.aes_encryptor.encrypt_message(encrypted_message)
                logger.debug("Applied AES encryption layer")
            
            return encrypted_message
            
        except Exception as e:
            logger.error(f"Hybrid encryption error: {e}")
            return message
    
    def decrypt_message(self, encrypted_message: str) -> str:
        """
        Decrypt message using hybrid approach (reverse order)
        
        Args:
            encrypted_message (str): Encrypted message
            
        Returns:
            str: Decrypted plain text message
        """
        try:
            decrypted_message = encrypted_message
            
            # First layer: AES decryption (reverse order)
            if self.use_aes and hasattr(self, 'aes_encryptor'):
                decrypted_message = self.aes_encryptor.decrypt_message(decrypted_message)
                logger.debug("Applied AES decryption layer")
            
            # Second layer: Tulu decryption
            if self.use_tulu:
                decrypted_message = self.tulu_encryptor.decrypt_message(decrypted_message)
                logger.debug("Applied Tulu decryption layer")
            
            return decrypted_message
            
        except Exception as e:
            logger.error(f"Hybrid decryption error: {e}")
            return encrypted_message

# Global instances for easy access
_tulu_encryptor = None
_hybrid_encryptor = None

def get_tulu_encryptor():
    """Get global Tulu encryptor instance"""
    global _tulu_encryptor
    if _tulu_encryptor is None:
        _tulu_encryptor = TuluEncryption()
    return _tulu_encryptor

def get_hybrid_encryptor():
    """Get global hybrid encryptor instance"""
    global _hybrid_encryptor
    if _hybrid_encryptor is None:
        _hybrid_encryptor = HybridEncryption()
    return _hybrid_encryptor

# Convenience functions
def tulu_encrypt(message: str) -> str:
    """Convenience function to encrypt with Tulu"""
    return get_tulu_encryptor().encrypt_message(message)

def tulu_decrypt(encrypted_message: str) -> str:
    """Convenience function to decrypt Tulu message"""
    return get_tulu_encryptor().decrypt_message(encrypted_message)

def hybrid_encrypt(message: str) -> str:
    """Convenience function for hybrid encryption"""
    return get_hybrid_encryptor().encrypt_message(message)

def hybrid_decrypt(encrypted_message: str) -> str:
    """Convenience function for hybrid decryption"""
    return get_hybrid_encryptor().decrypt_message(encrypted_message)

# Test and demonstration functions
if __name__ == "__main__":
    print("🔐 Testing Tulu Encryption System...")
    print("=" * 50)
    
    tulu_enc = TuluEncryption()
    hybrid_enc = HybridEncryption()
    
    test_messages = [
        "Hello World!",
        "This is a test message with numbers 12345",
        "Special characters: !@#$%^&*()",
        "Karnataka namaste! How are you?",
        "Mixed content: abc123!@# xyz789",
        "Unicode test: 🚀 Hello 世界 🌟",
        "Kannada: ನಮಸ್ತೆ ಕನ್ನಡ",
        ""  # Empty message test
    ]
    
    print("\n1. TULU ENCRYPTION TEST:")
    print("-" * 30)
    
    for i, msg in enumerate(test_messages):
        print(f"\nTest {i+1}:")
        print(f"Original:  '{msg}'")
        
        # Tulu encryption
        tulu_encrypted = tulu_enc.encrypt_message(msg)
        print(f"Tulu Enc:  '{tulu_encrypted[:50]}{'...' if len(tulu_encrypted) > 50 else ''}'")
        
        # Tulu decryption
        tulu_decrypted = tulu_enc.decrypt_message(tulu_encrypted)
        print(f"Tulu Dec:  '{tulu_decrypted}'")
        print(f"Match:     {msg == tulu_decrypted}")
        print(f"Is Tulu:   {tulu_enc.is_tulu_encrypted(tulu_encrypted)}")
    
    print("\n\n2. HYBRID ENCRYPTION TEST:")
    print("-" * 30)
    
    for i, msg in enumerate(test_messages[:3]):  # Test first 3 messages
        print(f"\nHybrid Test {i+1}:")
        print(f"Original:    '{msg}'")
        
        # Hybrid encryption (Tulu + AES)
        hybrid_encrypted = hybrid_enc.encrypt_message(msg)
        print(f"Hybrid Enc:  '{hybrid_encrypted[:50]}{'...' if len(hybrid_encrypted) > 50 else ''}'")
        
        # Hybrid decryption
        hybrid_decrypted = hybrid_enc.decrypt_message(hybrid_encrypted)
        print(f"Hybrid Dec:  '{hybrid_decrypted}'")
        print(f"Match:       {msg == hybrid_decrypted}")
    
    print("\n\n3. CHARACTER MAPPING DEMO:")
    print("-" * 30)
    
    demo_text = "namaste kannada"
    print(f"Demo text: '{demo_text}'")
    print("Character by character mapping:")
    
    result = ""
    i = 0
    while i < len(demo_text):
        if i < len(demo_text) - 1 and demo_text[i:i+2] in tulu_enc.tulu_mapping:
            two_char = demo_text[i:i+2]
            mapped = tulu_enc.tulu_mapping[two_char]
            print(f"  '{two_char}' → '{mapped}'")
            result += mapped
            i += 2
        elif demo_text[i] in tulu_enc.tulu_mapping:
            char = demo_text[i]
            mapped = tulu_enc.tulu_mapping[char]
            print(f"  '{char}' → '{mapped}'")
            result += mapped
            i += 1
        elif demo_text[i] in tulu_enc.symbol_mapping:
            char = demo_text[i]
            mapped = tulu_enc.symbol_mapping[char]
            print(f"  '{char}' → '{mapped}'")
            result += mapped
            i += 1
        else:
            scrambled = tulu_enc.scramble_text(demo_text[i])
            print(f"  '{demo_text[i]}' → '{scrambled}' (scrambled)")
            result += scrambled
            i += 1
    
    print(f"Final result: '{result}'")
    
    print("\n✅ All tests completed!")
    print("\nEncryption layers available:")
    print("- 🔤 Tulu character mapping")
    print("- 🔢 Symbol and number mapping")
    print("- 🌀 Golden ratio scrambling")
    print("- 🔐 AES-256 encryption (if available)")
    print("\nHybrid system provides multiple layers of security! 🚀")